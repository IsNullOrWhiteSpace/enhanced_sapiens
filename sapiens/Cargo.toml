[package]
name = "sapiens"
version = "0.10.2"
edition.workspace = true
authors.workspace = true
description = "Core - Sapiens"
repository.workspace = true
license.workspace = true
publish.workspace = true
readme = "../README.md"
keywords.workspace = true
categories.workspace = true
documentation = "https://docs.rs/sapiens"

[features]
default = ["clap"]

# derive Clap traits
clap = ["dep:clap"]

[dependencies]
tokio = { version = "1.34.0" }
tracing = "0.1.40"
async-trait = "0.1.74"
lazy_static = "1.4.0"

serde = { version = "1.0.193", features = ["derive"] }
serde_yaml = "0.9.27"

clap = { version = "4.5.1", optional = true }

# OpenAI API - OpenAI and lm-sys/FastChat
async-openai = "0.16.3"
tokenizers = { version = "0.15.0", features = [] }

# GCP Vertex AI Generative Language Models
gcp-vertex-ai-generative-language = "0.1.2"

thiserror = "1.0.50"

[dev-dependencies]
indoc = "2"
insta = { version = "1.34.0", features = ["yaml"] }
