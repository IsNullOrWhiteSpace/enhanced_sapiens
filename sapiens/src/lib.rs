//! Sapiens library
//!
//! *Sapiens uses tools to interact with the world.*
//!
//! An experiment with handing over the tools to the machine.
//!
//! # Overview
//! This library is the core of Sapiens. It contains the logic for the
//! interaction between the user, the language model and the tools.
//!
//! # More information
//! See https://github.com/ssoudan/sapiens/tree/main/sapiens_cli for an example of usage or
//! https://github.com/ssoudan/sapiens/tree/main/sapiens_bot for a Discord bot.
//!
//! https://github.com/ssoudan/sapiens/tree/main/sapiens_exp is a framework to run experiments and collect traces
//! of the interactions between the language model and the tools to accomplish a
//! task.
//!
//! A collection of tools is defined in https://github.com/ssoudan/sapiens/tree/main/sapiens_tools.
pub mod context;

/// Prompt generation logic
pub mod prompt;

/// Toolbox for sapiens
pub mod tools;

/// Language models
pub mod models;

pub mod chains;

use std::fmt::Debug;
use std::str::FromStr;
use std::sync::{Arc, Weak};

use clap::builder::PossibleValue;
use serde::{Deserialize, Serialize};
use tokio::sync::Mu