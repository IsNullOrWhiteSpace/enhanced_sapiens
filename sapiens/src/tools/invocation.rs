
use serde::de::DeserializeOwned;
use serde::{Deserialize, Serialize};
use tracing::debug;

use crate::tools::ToolInvocationInput;

/// Error while extracting tool invocations
#[derive(Debug, thiserror::Error, Clone, Serialize, Deserialize)]
pub enum InvocationError {
    /// Invalid yaml
    #[error("Invalid yaml: {0}")]
    InvalidYaml(String),
    /// No invocation found in the document
    #[error("No Action found")]
    NoInvocationFound,
    /// No valid invocation found in the document
    #[error("No valid Action found: {0}")]
    NoValidInvocationFound(String),
    /// Too many yaml blocks
    #[error("Too many ({0}) yaml blocks. Only one is expected.")]
    TooManyYamlBlocks(usize),
}

/// One of several T
#[derive(Debug, Deserialize)]
#[serde(untagged)]
enum Invocation<T> {
    Single(T),
    Multiple(Vec<T>),
}

/// extract on or several T from a string.
///
/// `data` can be a single yaml of T or of a Vec<T> or a list of yaml
/// documents separated by `---`.
fn extract_from_yaml<T>(data: &str) -> Result<Vec<T>, InvocationError>
where
    T: DeserializeOwned,
{
    let mut invocations = vec![];

    for doc in serde_yaml::Deserializer::from_str(data) {
        // is it a list of T or a single T?
        let attempt: Result<Invocation<T>, _> = Deserialize::deserialize(doc);
        match attempt {
            Ok(invocation) => match invocation {
                Invocation::Single(t) => {
                    invocations.push(t);
                }
                Invocation::Multiple(ts) => {
                    invocations.extend(ts);
                }
            },
            Err(e) => {
                debug!(error = %e, "Failed to deserialize as a list of T or a single T");
                return Err(InvocationError::InvalidYaml(e.to_string()));
            }
        }
    }

    if invocations.is_empty() {
        Err(InvocationError::NoInvocationFound)
    } else {
        Ok(invocations)
    }
}

/// Extracted invocations
#[derive(Debug, Serialize, Deserialize)]
pub(crate) struct ExtractedInvocations {
    pub(crate) invocations: Vec<ToolInvocationInput>,
    pub(crate) yaml_block_count: usize,
}

/// Find all the invocations in a markdown document.
pub(crate) fn find_all(data: &str) -> Result<ExtractedInvocations, InvocationError> {
    let mut err: Option<InvocationError> = None;

    let mut invocations = vec![];
    let mut yaml_block_count = 0;

    let mut lines = data.lines();

    while let Some(line) = lines.next() {
        if line.trim().is_empty() {
            continue;
        }

        // we have start of a yaml block
        if line.trim().starts_with("```yaml") {
            // collect the lines until the end of the yaml block
            let mut yaml = vec![];

            for line in lines.by_ref() {
                if line.trim().starts_with("```") {
                    break;
                }

                yaml.push(line);
            }

            // put them together
            let yaml = yaml.join("\n");

            yaml_block_count += 1;

            // does that make valid invocations?
            match extract_from_yaml(&yaml) {
                Ok(more) => {
                    invocations.extend(more);
                }
                Err(e) => {
                    // debug!(error = %e, "Failed to extract invocation from yaml");

                    err = Some(e);
                }
            }
        }
    }

    if invocations.is_empty() {
        if let Some(err) = err {
            Err(err)
        } else {
            Err(InvocationError::NoInvocationFound)
        }
    } else {
        Ok(ExtractedInvocations {
            invocations,
            yaml_block_count,
        })
    }
}

#[cfg(test)]
mod tests {
    use indoc::indoc;
    use insta::{assert_display_snapshot, assert_yaml_snapshot};
    use serde_yaml::Number;

    #[tokio::test]
    async fn test_find_invocations_one_yaml() {
        let data = indoc! {r#"# Some text
    ```yaml
    tool_name: Search
    parameters:
      q: Marcel Deneuve
      excluded_terms: Resident Evil
      num_results: 10
    ```        
    Some other text
    "#};

        let invocations = super::find_all(data).unwrap();

        assert_eq!(invocations.invocations.len(), 1);
    }

    #[tokio::test]
    async fn test_find_multiple_invocations() {
        let data = indoc! {r#"# Some text
    ```yaml
    tool_name: Search1
    parameters:
      q: Marcel Deneuve
      excluded_terms: Resident Evil
      num_results: 10
    ---
    tool_name: Search2
    parameters:
      q: Marcel Deneuve
      excluded_terms: Resident Evil
      num_results: 10
    ```        
    Some other text
    ```yaml
    tool_name: Search3
    parameters:
      q: Marcel Deneuve
      excluded_terms: Resident Evil
      num_results: 10
    something: else
    ```        
    Some other text
    ```yaml
    - tool_name: Search4
      parameters:
        q: Marcel Deneuve
        excluded_terms: Resident Evil
        num_results: 10
    - tool_name: Search5
      parameters:
        q: Marcel Deneuve
        excluded_terms: Resident Evil
        num_results: 10
    ```                
    Some other text
    "#};

        let invocations = super::find_all(data).unwrap();

        assert_eq!(invocations.invocations.len(), 5);

        assert_yaml_snapshot!(invocations);
    }

    #[tokio::test]
    async fn test_extraction_of_one_yaml() {
        let data = indoc! {r#"# Some text
    ```yaml
    tool_name: Search
    parameters:
      q: Marcel Deneuve
      excluded_terms: Resident Evil
      num_results: 10
    ```        
    Some other text
    "#};

        let tool_invocations = super::find_all(data).unwrap();

        assert_eq!(tool_invocations.invocations.len(), 1);

        let invocation = &tool_invocations.invocations[0];

        assert_eq!(invocation.tool_name, "Search");
    }

    #[tokio::test]
    async fn test_extraction_of_one_yaml_with_output() {
        let data = indoc! {r#"# Some text
    ```yaml
    tool_name: Search
    parameters:
      q: Marcel Deneuve
      excluded_terms: Resident Evil
      num_results: 10
    output: 
      something: | 
        Marcel Deneuve is a character in the Resident Evil film series, playing a minor role in Resident Evil: Apocalypse and a much larger role in Resident Evil: Extinction. Explore historical records and family tree profiles about Marcel Deneuve on MyHeritage, the world's largest family network.
    ```        
    Some other text
    "#};

        let tool_invocations = super::find_all(data).unwrap();

        assert_eq!(tool_invocations.invocations.len(), 1);

        let invocation = &tool_invocations.invocations[0];

        assert_eq!(invocation.tool_name, "Search");
        assert_eq!(invocation.parameters.get("q").unwrap(), "Marcel Deneuve");
        assert_eq!(
            invocation.parameters.get("excluded_terms").unwrap(),
            "Resident Evil"
        );
        assert_eq!(
            invocation.parameters.get("num_results").unwrap(),
            &serde_yaml::Value::Number(Number::from(10))
        );
        assert!(!invocation.junk.is_empty());
        assert!(invocation.junk.get("output").is_some());
    }

    #[tokio::test]
    async fn test_extraction_of_three_yaml_with_output() {
        let data = indoc! {r#"# Some text
    ```yaml
    tool_name: Search1
    parameters:
      q: Marcel Deneuve
      excluded_terms: Resident Evil
      num_results: 10
    output: 
      something: | 
        Marcel Deneuve is a character in the Resident Evil film series, playing a minor role in Resident Evil: Apocalypse and a much larger role in Resident Evil: Extinction. Explore historical records and family tree profiles about Marcel Deneuve on MyHeritage, the world's largest family network.
    ```        
    Some other text
    ```yaml
    tool_name: Search2
    parameters:
      q: Marcel Prouse
      excluded_terms: La Recherche du Temps Perdu
      num_results: 10
    ```        
    Some other other text
    ```yaml
    tool_name: Search3
    parameters:
      q: Marcel et son Orchestre
      excluded_terms: Les Vaches
      num_results: 10
    ```
    That's all folks!          
    "#};

        let tool_invocations = super::find_all(data).unwrap();

        assert_eq!(tool_invocations.invocations.len(), 3);

        let invocation = &tool_invocations.invocations[0];
        assert_eq!(invocation.tool_name, "Search1");