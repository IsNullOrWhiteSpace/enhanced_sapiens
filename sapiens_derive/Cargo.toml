
[package]
name = "sapiens_derive"
version= "0.4.4"
edition.workspace = true
authors.workspace = true
description = "The macro - a simple bot to conquer the world"
repository.workspace = true
license.workspace = true
publish.workspace = true
readme = "../README.md"
keywords.workspace = true
categories.workspace = true
documentation = "https://docs.rs/sapiens_derive"

[dependencies]
proc-macro2 = "1.0"
quote = "1.0"
syn = { version = "2.0" }
darling = "0.20.3"
darling_macro = "0.20.3"

[lib]
proc-macro = true