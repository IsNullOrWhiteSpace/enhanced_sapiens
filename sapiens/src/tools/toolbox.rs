
use std::collections::HashMap;
use std::fmt::Debug;
use std::sync::Arc;

use serde::{Deserialize, Serialize};
use tokio::sync::RwLock;
use tracing::{debug, info};

use crate::tools;
use crate::tools::invocation::InvocationError;
use crate::tools::{
    AdvancedTool, TerminalTool, TerminationMessage, Tool, ToolDescription, ToolUseError,
};