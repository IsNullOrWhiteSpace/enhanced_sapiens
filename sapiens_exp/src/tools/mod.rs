
//! Bake the acceptance test for the task in the tools
//! Scenario -> (Recipe, Config)
//!
//! Recipe -> (Task, Acceptance criteria)
//!
//! Acceptance criteria -> Tools
//!
//! run_until_completion: (Task, Config, Tools) -> (Trace, Stats, Acceptance
//! results)
//!
//! evaluate::Trial::analyze: (Trace, Stats, Acceptance results) -> Analysis
//!
//! evaluate::Trial::build: (Trace, State, Acceptance results, Task, Analysis)
//! -> Trial

use std::marker::PhantomData;
use std::ops::DerefMut;
use std::sync::Arc;

use sapiens::tools::{Describe, ProtoToolDescribe, ProtoToolInvoke, ToolDescription, ToolUseError};
use serde::{Deserialize, Serialize};
use tokio::sync::Mutex;

/// Scenario 0: Preparing a cereal bowl
pub mod scenario_0;

mod scenario_1;

/// The state of a scenario
pub trait State: Send + Sync {
    /// Reset the state to its initial state
    fn reset(&mut self);
