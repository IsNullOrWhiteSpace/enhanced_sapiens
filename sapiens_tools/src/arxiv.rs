use std::fmt::Display;

use arxiv;
use arxiv::{Arxiv, ArxivQuery};
use sapiens::tools::{Describe, ProtoToolDescribe, ProtoToolInvoke, ToolDescription, ToolUseError};
use sapiens_derive::{Describe, ProtoToolDescribe, ProtoToolInvoke};
use serde::{Deserialize, Serialize};

/// A Tool to query arXiv.
///
/// arXiv is a free distribution service and an open-access archive for
/// millions scholarly articles in the fields of physics, mathematics, computer
/// science, quantitative biology, quantitative finance, statistics, electrical
/// engineering and systems science, and economics. Materials on this site are
/// not peer-reviewed by arXiv.
#[derive(Debug, ProtoToolInvoke, ProtoToolDescribe)]
#[tool(name = "Arxiv", input = "ArxivToolInput", output = "ArxivToolOutput")]
pub struct ArxivTool {}

/// Sort order
#[derive(Debug, Deserialize, Serialize, Default, Clone)]
pub enum SortOrder {
    /// Ascending
    #[serde(rename = "ascending")]
    Ascending,
    /// Descending
    #[serde(rename = "descending")]
    #[default]
    Descending,
}

impl Display for SortOrder {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            SortOrder::Ascending => write!(f, "ascending"),
            SortOrder::Descending => write!(f, "descending"),
        }
    }
}

/// Sort by
#[derive(Debug, Deserialize, Serialize, Default, Clone)]
pub enum SortBy {
    /// Relevance
    #[serde(rename = "relevance")]
    #[default]
    Relevance,
    /// Last updated date
    #[serde(rename = "lastUpdatedDate")]
    LastUpdatedDate,
    /// Submitted date
    #[serde(rename = "submittedDate")]
    SubmittedDate,
}

impl Display for SortBy {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            SortBy::Relevance => write!(f, "relevance"),
            SortBy::LastUpdatedDate => write!(f, "lastUpdatedDate"),
            SortBy::SubmittedDate => write!(f, "submittedDate"),
        }
    }
}

/// [`ArxivTool`] input
///
/// Arxiv API documentation query specification
#[derive(Debug, Deserialize, Serialize, Describe)]
pub struct ArxivToolInput {
    /// search_query: Search query - see https://info.arxiv.org/help/api/user-manual.html
    /// for details. E.g. `cs.AI` or `cat:cs.AI` or `au:John Smith`
    /// The fields that can be searched are: `ti` (title), `au` (author), `abs`
    /// (abstract), `co` (comment), `jr` (journal reference), `cat` (subject
    /// category), `rn` (report number), `id` (id (use id_list instead)),
    /// `all` (all of the above). Operators: `AND`, `OR`, `ANDNOT`.
    /// You cannot search on publication or last update date.
    pub search_query: String,

    /// id_list: Comma-separated list of arXiv IDs to return
    pub id_list: Option<String>,

    /// start: Result offset for pagination
    pub start: Option<i32>,

    /// max_results: Maximum number of results to return in a single response.
    /// Default is 10. Maximum allowed value is 100.
    pub max_results: Option<i32>,

    /// Sort by. Can be either `relevance`, `lastUpdatedDate` or
    /// `submittedDate`. Default is `relevance`.
    pub sort_by: Option<SortBy>,

    /// Sort order. Can be either `ascending` or `descending`.
    /// Default is `descending`.
    pub sort_order: Option<SortOrder>,

    /// True to gather PDF url - default is false
    pub show_pdf_url: Option<bool>,

    /// True to gather authors - default is false
    pub show_authors: Option<bool>,

    /// True to gather comments - default is false
    pub show_comments: Option<bool>,

    /// True to gather summary - default is false
    pub show_summary: Option<bool>,
}

impl From<&ArxivToolInput> for ArxivQuery {
    fn from(input: &ArxivToolInput) -> Self {
        ArxivQuery {
            base_url: "https://export.arxiv.org/api/query?".to_string(),
            search_query: input.search_query.clone(),
            id_list: input.id_list.clone().unwrap_or_default(),
            start: input.start,
            max_results: input.max_results,
            sort_by: input.sort_by.clone().unwrap_or_default().to_string(),
            sort_order: input.sort_order.clone().unwrap_or_default().to_string(),
        }
    }
}

/// [`ArxivTool`] output
#[derive(Debug, Deserialize, Serialize, Describe)]
pub struct ArxivToolOutput {
    // FUTURE(ssoudan) proc_macro_derive to generate this
    /// query result. `ArxivResult` is an object containing the following
    /// fields:
    /// - `id`: <str> arXiv ID
    /// - `updated`: <str> last updated date
    /// - `published`: <str> published date
    /// - `title`: <str> title
    /// - `summary`: <Optional[str]> summary - omitted unless `show_summary` is
    ///   true - can be quite long
    /// - `authors`: <list[str]> authors - omitted unless `show_authors` is true
    /// - `pdf_url`: <Optional[str]> PDF URL - omitted unless `show_pdf_url` is
    ///   true
    /// - `comments`: <Optional[str]> Comments - omitted unless `show_comments`
    ///   is true
    result: Vec<ArxivResult>,
}

/// Arxiv result
#[derive(Debug, Deserialize, Serialize, Describe)]
pub struct ArxivResult {
    /// arXiv ID
    pub id: String,
    /// last updated date
    pub updated: String,
    /// published date
    pub published: String,
    /// title
    pub title: String,
    /// summary - only if `show_summary` is true - can be quite long
    #[serde(skip_serializing_if = "Option::is_none")]
    pub summary: Option<String>,
    /// authors - only if `show_authors` is true
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub authors: Vec<String>,
    /// PDF URL - only if `show_pdf_url` is true
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pdf_url: Option<String>,
    /// Comments - only if `show_comments` is true
    #[serde(skip_serializing_if = "Option::is_none")]
    pub comment: Option<String>,
}

impl From<Arxiv> for ArxivResult {
    fn from(arxiv: Arxiv) -> Self {
        ArxivResult {
            id: arxiv.id,
            updated: arxiv.updated,
            published: arxiv.published,
            title: arxiv.title,
            summary: Some(arxiv.summary),
            authors: arxiv.authors,
            pdf_url: Some(arxiv.pdf_url),
            comment: arxiv.comment,
        }
    }
}

impl ArxivTool {
    /// Create a new [`ArxivTool`]
    pub async fn new() -> ArxivTool {
        ArxivTool {}
    }

    #[tracing::instrument(skip(self))]
    async fn invoke_typed(&self, input: &ArxivToolInput) -> Result<ArxivToolOutput, ToolUseError> {
        let query = ArxivQuery::from(input);

        if query.max_results.unwrap_or(0) > 100 {
            return Err(ToolUseError::InvocationFailed(
                "max_results cannot be greater than 100".to_string(),
            ));
        }
