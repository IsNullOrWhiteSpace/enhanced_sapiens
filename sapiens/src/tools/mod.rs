
use std::collections::HashMap;
use std::fmt::Debug;

use serde::ser::SerializeMap;
use serde::{Deserialize, Serialize, Serializer};
use toolbox::Toolbox;
use tracing::warn;

use crate::tools::invocation::{ExtractedInvocations, InvocationError};

/// Tools to extract Tool invocations from a messages
pub mod invocation;

/// Collection of tools
pub mod toolbox;

/// Part of a [`Format`]
#[derive(Debug, Clone)]
pub struct FieldFormat {
    /// Name of the field
    pub name: String,
    /// Type of the field
    pub r#type: String,
    /// True if the field is optional
    pub optional: bool,
    /// Description of the field
    pub description: String,
}

/// Input or output format of a tool
pub trait Describe {
    /// Describe the format
    fn describe() -> Format;
}

/// Format of [`Tools`] input and output
#[derive(Debug, Clone, Default)]
pub struct Format {
    /// Fields of the format
    pub fields: Vec<FieldFormat>,
}

impl Serialize for Format {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let n = self.fields.len();
        let mut map = serializer.serialize_map(Some(n))?;
        for field in &self.fields {
            let description = if field.optional {
                format!("<{}> {} (optional)", field.r#type, field.description)
            } else {
                format!("<{}> {}", field.r#type, field.description)
            };
            map.serialize_entry(&field.name, &description)?;
        }
        map.end()
    }
}

impl From<Vec<FieldFormat>> for Format {
    fn from(fields: Vec<FieldFormat>) -> Self {
        Format { fields }
    }
}

/// Tool description
#[derive(Debug, Serialize, Clone)]
pub struct ToolDescription {
    /// Name of the tool
    pub name: String,
    /// Description of the tool
    pub description: String,
    /// Input format
    pub parameters: Format,
    /// Output format
    pub responses_content: Format,
}

impl ToolDescription {
    /// Create a new tool description
    pub fn new(
        name: &str,
        description: &str,
        parameters: Format,
        responses_content: Format,
    ) -> Self {
        ToolDescription {
            name: name.to_string(),
            description: description.to_string(),
            parameters,
            responses_content,
        }
    }
}

/// Error while using a tool
#[derive(Debug, thiserror::Error, Clone, Serialize, Deserialize)]
pub enum ToolUseError {
    /// Tool not found
    #[error("Tool not found: {0}")]
    ToolNotFound(String),
    /// Tool invocation failed
    #[error("Tool invocation failed: {0}")]
    InvocationFailed(String),
    /// Failed to serialize the output
    #[error("Failed to serialize the output: {0}")]
    InvalidOutput(String),
    /// Failed to deserialize the input
    #[error("Failed to deserialize the parameters: {0}")]
    InvalidInput(String),
}

/// A tool invocation input
#[derive(Serialize, Deserialize, Debug)]
pub(crate) struct ToolInvocationInput {
    /// The tool to invoke
    tool_name: String,
    // FUTURE(ssoudan) should this be flattened?
    // FUTURE(ssoudan) should this be called `spec` or `arguments` or `parameters`?
    /// The input to the tool
    parameters: serde_yaml::Value,
    /// The junk
    #[serde(skip_serializing_if = "HashMap::is_empty", flatten)]
    junk: HashMap<String, serde_yaml::Value>,
}

/// Something meant to become a [`Tool`] - description
pub trait ProtoToolDescribe {
    /// the description of the tool
    fn description(&self) -> ToolDescription;
}

/// Something meant to become a [`Tool`] - invocation
#[async_trait::async_trait]
pub trait ProtoToolInvoke {
    /// Invoke the tool
    async fn invoke(&self, input: serde_yaml::Value) -> Result<serde_yaml::Value, ToolUseError>;
}

/// A Tool - the most basic kind of tools. See [`AdvancedTool`] and
/// [`TerminalTool`] for more.
#[async_trait::async_trait]
pub trait Tool: Sync + Send {
    /// the description of the tool
    fn description(&self) -> ToolDescription;

    /// Invoke the tool
    // FUTURE(ssoudan) Box<Deserialize>?
    async fn invoke(&self, input: serde_yaml::Value) -> Result<serde_yaml::Value, ToolUseError>;
}

#[async_trait::async_trait]
impl<T: Sync + Send> Tool for T
where
    T: ProtoToolDescribe + ProtoToolInvoke,
{
    fn description(&self) -> ToolDescription {
        self.description()
    }

    async fn invoke(&self, input: serde_yaml::Value) -> Result<serde_yaml::Value, ToolUseError> {
        self.invoke(input).await
    }
}

/// A termination message
///
/// This is the message that is sent to the user when a chain of exchanges
/// terminates.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TerminationMessage {
    /// The final textual answer for this task.
    pub conclusion: String,
    /// The original question that was asked to the user.
    pub original_question: String,
}

/// A [`Tool`] that wraps a chain of exchanges
#[async_trait::async_trait]
pub trait TerminalTool: Tool + Sync + Send {
    /// done flag.
    async fn is_done(&self) -> bool {
        false
    }

    /// Take the done flag.
    async fn take_done(&self) -> Option<TerminationMessage> {
        None
    }
}

/// A [`Tool`]  that can benefit from a [`Toolbox`]
#[async_trait::async_trait]
pub trait AdvancedTool: Tool {
    /// Invoke the tool with a [`Toolbox`]
    async fn invoke_with_toolbox(
        &self,
        toolbox: Toolbox,
        input: serde_yaml::Value,
    ) -> Result<serde_yaml::Value, ToolUseError>;
}

async fn choose_invocation(
    tool_invocations: ExtractedInvocations,
) -> Result<ToolInvocationInput, InvocationError> {
    // TODO(ssoudan) customizable level of strictness
    if tool_invocations.yaml_block_count > 1 {
        return Err(InvocationError::TooManyYamlBlocks(
            tool_invocations.yaml_block_count,
        ));
    }

    // if no tool_invocations are found, we return an error
    if tool_invocations.invocations.is_empty() {
        return Err(InvocationError::NoInvocationFound);
    }

    // We just take the first one
    let mut invocation = tool_invocations.invocations.into_iter().next().unwrap();

    // FUTURE(ssoudan) clean up the object and return this one instead
    // if any tool_invocations have an 'output' field, we return an error
    if !invocation.junk.is_empty() {
        let junk_keys = invocation
            .junk
            .keys()
            .cloned()
            .collect::<Vec<String>>()
            .join(", ");

        warn!(
            ?junk_keys,
            "The Action should not have fields: {}.", junk_keys
        );

        // We just remove them for now
        invocation.junk.clear();

        // return Err(InvocationError::NoValidInvocationFound(format!(
        //     "The Action cannot have fields: {}. Only `command` and `input`
        // are allowed.",     junk_keys
        // )));
    }

    Ok(invocation)
}

#[cfg(test)]
mod tests {
    use std::collections::HashMap;

    use insta::assert_display_snapshot;
    use serde::{Deserialize, Serialize};

    #[derive(Debug, Serialize, Deserialize)]
    struct FakeToolInput {
        q: String,
        excluded_terms: Option<String>,
        num_results: Option<u32>,
    }

    #[derive(Debug, Serialize, Deserialize)]
    struct FakeToolOutput {
        items: Vec<String>,
    }

    #[tokio::test]
    async fn test_serializing_tool_invocation() {
        let input = FakeToolInput {
            q: "Marcel Deneuve".to_string(),
            excluded_terms: Some("Resident Evil".to_string()),
            num_results: Some(10),
        };

        let output = FakeToolOutput {
        items: vec![
            "Marcel Deneuve is a character in the Resident Evil film series,".to_string(), 
            "playing a minor role in Resident Evil: Apocalypse and a much larger".to_string(),
            " role in Resident Evil: Extinction. Explore historical records and ".to_string(),
            "family tree profiles about Marcel Deneuve on MyHeritage, the world's largest family network.".to_string()
        ]

        };

        let junk = vec![("output".to_string(), serde_yaml::to_value(output).unwrap())];

        let invocation = super::ToolInvocationInput {
            tool_name: "Search".to_string(),
            parameters: serde_yaml::to_value(input).unwrap(),
            junk: HashMap::from_iter(junk.into_iter()),
        };

        let serialized = serde_yaml::to_string(&invocation).unwrap();

        assert_display_snapshot!(serialized);
    }
}