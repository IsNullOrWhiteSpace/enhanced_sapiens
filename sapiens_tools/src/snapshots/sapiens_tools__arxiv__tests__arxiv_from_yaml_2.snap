---
source: sapiens_tools/src/arxiv.rs
expression: input
---
search_query: "cat:cs.DB"
id_list: ~
start: ~
max_results: 4
sort_by: ~
sort_order: ~
show_pdf_url: true
show_authors: true
show_comments: ~
show_summary: ~

